Add RFC-006: RBAC - Supporting DocumentDB's RBAC model on PG

nitinahuja89 opened the pull request:
RFC for supporting DocumentDB's RBAC model on PG

Closes https://github.com/documentdb/documentdb/issues/378

xgerman commented on rfcs/006-rbac-support.md:
```diff
@@ -0,0 +1,1246 @@
+
+# RFC-006: RBAC: Supporting DocumentDB's RBAC model on PG
+
+**Note**: Throughout this document, "PG" refers to PostgreSQL.
```
PostgreSQL - only one S

nitinahuja89 replied:
There is only one S.

xgerman commented on rfcs/006-rbac-support.md:
```diff
@@ -0,0 +1,1246 @@
+
+**Success criteria:**
+
+- The architecture provides a way for DocumentDB to support all standard MongoDB privileges, including those not natively available in PostgreSQL
```
This is debatable as a goal. We might choose to only support a subste to make it easier to map to PG model.

nitinahuja89 replied:
If we only use PG privileges for RBAC we will not be able to have compatibility with MongoDB RBAC model

xgerman commented on rfcs/006-rbac-support.md:
```diff
@@ -0,0 +1,1246 @@
+  - Privilege enforcement is consistent with MongoDB behavior
+- Performance overhead of privilege checking is minimal and acceptable for production use
+- Provides an easy mechanism for contributors to add support for new privileges when implementing features that require corresponding privilege checks
+
```
Add security as an additional criteria

nitinahuja89 replied:
Do you have a proposal of how you would like security aspect to be worded?

nitinahuja89 replied:
Also, this is a security feature so it's kind of implicit that security is the primary aspect here, but happy to add it explicitly if you have a specific wording suggestion.

xgerman commented on rfcs/006-rbac-support.md:
```diff
@@ -0,0 +1,1246 @@
+- More invasive - need to add the check in all command handlers
+
+#### Option C: PG Hooks
+
```
we found that hooks occasionally get skipped. So this would need to be prototyped to completely understand the implications to perfroemance when hook skipping is disabled, etc.

nitinahuja89 replied:
I didn't see the hooks getting skipped - so we could dive into the cases where you do see it skipped. But in any case this is not the recommended approach so we probably don't need to dive deep into this.

visridha commented on rfcs/006-rbac-support.md:
```diff
@@ -0,0 +1,1246 @@
+- Supports BSON-specific types (ObjectId, etc.)
+
+**Cons**:
+- Less type safety - validation must be done in application code
```
FWIW this could be enforced a bit more via jsonschema or CHECK constraints right?

nitinahuja89 replied:
Yes that's right.

visridha commented on rfcs/006-rbac-support.md:
```diff
@@ -0,0 +1,1246 @@
+- Every update to user/role information (createUser, grantRole, dropRole, etc.) increments the version number in the same transaction
+- Each cache stores the version number it was built from
+- Before using cache, compare cached version with current version in metadata_version table
+- If versions differ, invalidate cache and rebuild from current metadata
```
Presume We'll borrow stuff from the catalog cache in PG for inval right?

nitinahuja89 replied:
Yes. Sean is working on the low level design which will have more details about the invalidation options.

nitinahuja89 committed:
Add RFC-006: RBAC - Supporting DocumentDB's RBAC model on PG

This RFC proposes an architecture for implementing MongoDB-compatible RBAC
on PostgreSQL, addressing the challenge that MongoDB's 110+ granular
privileges don't all have direct equivalents in PostgreSQL's native
permission system.

The RFC explores two design dimensions:
1. Privilege Storage: Normalized metadata tables vs pgbson documents vs
   two-tier hybrid system
2. Privilege Enforcement: Gateway layer vs extension handlers vs PG hooks
   vs hybrid enforcement

Recommended approach:
- Use pgbson document metadata for flexible privilege storage
- Implement custom authorization in extension command handlers
- Maintain PG ACL synchronization for CRUD operations to enable legitimate
  direct SQL access for administrative operations and analytics
- Use ALTER DEFAULT PRIVILEGES with schema-scoped owner roles for efficient
  new object handling

Implementation will be phased:
- Phase 1: Build custom authorization infrastructure (temporarily use PG
  ACLs for CRUD operations)
- Phase 2: Migrate to unified custom authorization once validated

The RFC includes comprehensive appendices covering MongoDB's complete
privilege list and detailed SQL patterns for PG ACL synchronization.

Signed-off-by: Nitin Ahuja <nitinahuja.89@gmail.com>

xgerman commented on rfcs/006-rbac-support.md:
```diff
@@ -0,0 +1,1246 @@
+   - Many diagnostic and session management commands could partially map to PG system catalog queries and functions:
+     - `collStats` → SELECT on `pg_stat_user_tables`, `pg_class`, `pg_table_size()` functions
+     - `dbStats` → SELECT on `pg_database`, `pg_stat_database`, `pg_database_size()` functions  
+     - `listDatabases` → SELECT on `pg_database` catalog
```
you would need to map this to our documentdb metadata tables, e.g. documentdb_api_catalog.collections

xgerman commented on rfcs/006-rbac-support.md:
```diff
@@ -0,0 +1,1246 @@
+     - `dbStats` → SELECT on `pg_database`, `pg_stat_database`, `pg_database_size()` functions  
+     - `listDatabases` → SELECT on `pg_database` catalog
+     - `listCollections` → SELECT on `pg_tables` or `pg_class`
+     - `listIndexes` → SELECT on `pg_indexes` or `pg_stat_user_indexes`
```
documentdb_api_catalog.collection_indexes

xgerman commented on rfcs/006-rbac-support.md:
```diff
@@ -0,0 +1,1246 @@
+     - `listDatabases` → SELECT on `pg_database` catalog
+     - `listCollections` → SELECT on `pg_tables` or `pg_class`
+     - `listIndexes` → SELECT on `pg_indexes` or `pg_stat_user_indexes`
+     - `serverStatus` → SELECT on various `pg_stat_*` views
```
that seems also needs to consider statis of "conencted" servers i replica_set so would need some replication slot logic

xgerman commented on rfcs/006-rbac-support.md:
```diff
@@ -0,0 +1,1246 @@
+     - `listIndexes` → SELECT on `pg_indexes` or `pg_stat_user_indexes`
+     - `serverStatus` → SELECT on various `pg_stat_*` views
+     - `inprog` → SELECT on `pg_stat_activity` view
+     - `killop` → EXECUTE on `pg_terminate_backend()` and `pg_cancel_backend()` functions
```
refer to killop in  RFC-0015: Logical Session Management  -- right now that won't be backed by a DB table but that should happen in a multi-replica case...

xgerman commented on rfcs/006-rbac-support.md:
```diff
@@ -0,0 +1,1246 @@
+     - `inprog` → SELECT on `pg_stat_activity` view
+     - `killop` → EXECUTE on `pg_terminate_backend()` and `pg_cancel_backend()` functions
+     - `top` → SELECT on `pg_stat_statements` extension tables, `pg_stat_activity` view
+     - `listSessions` → SELECT on `pg_stat_activity` view (with appropriate filters)
```
see  RFC-0015: Logical Session Management

xgerman commented on rfcs/006-rbac-support.md:
```diff
@@ -0,0 +1,1246 @@
+     - PG's SELECT privilege on `pg_stat_activity` is all-or-nothing - you either see all sessions or none (can't restrict to specific databases)
+     - MongoDB privileges like `killop` allow killing own operations by default, but `killop` as a granted privilege allows killing any operation - PG doesn't have this nuanced distinction
+
+4. **Authentication and Security** - `changePassword`, `changeOwnPassword`, `changeCustomData`, `changeOwnCustomData`, `setAuthenticationRestriction`, etc.
```
also there is a complication if we use a 3rd party auth system

xgerman commented on rfcs/006-rbac-support.md:
```diff
@@ -0,0 +1,1246 @@
+   - MongoDB has a built-in profiling system (`enableProfiler`) to track slow queries and operations - PG has logging and statistics extensions but no equivalent profiling privilege model
+   - Query optimization privileges (`planCacheRead`, `planCacheWrite`, `planCacheIndexFilter`, `querySettings`) control access to query plan caching and optimization features - PG's query planner is not access-controlled at this granularity
+   - `cpuProfiler` privilege for performance analysis - PG has different performance monitoring tools (pg_stat_statements, perf, etc.) without privilege-level access control
+   - These features are critical for database performance tuning but represent distinctly different models between MongoDB and PG
```
Comment: given that we are a different database architecture it might be ok to have different tuning and profiling semantics... not sure if a mongo profiling tool would help a lot in our setup.

xgerman reviewed (commented):
Missing: Privilege escalation prevention

The RFC doesn't discuss how to prevent privilege escalation. In MongoDB, a user with grantRole privilege can only grant roles they
already possess (or subordinate roles). Without this constraint, a user with grantRole on database "sales" could grant themselves dbAdmin
or even escalate to broader privileges.

This is a security-critical design gap. The RFC should describe:

 - What constraints apply when granting/revoking roles
 - How role hierarchy cycles are prevented
 - How grantRole + revokeRole privileges interact with the role hierarchy


Cache invalidation: serialization bottleneck

The metadata_version single-row table (line 370) becomes a serialization point for all concurrent role/privilege mutations. In a
multi-tenant environment with frequent grantRole/revokeRole calls, this could cause contention. Consider:

 - Using PG's native catalog invalidation hooks (as visridha suggested, line 369, and Nitin confirmed Sean is working on the LLD)
 - Advisory locks or notification channels instead of version polling
